flag_set.go

```
@@ -264,19 +256,23 @@ func (fs *FlagSet) parseLongFlag(arg string, args []string) ([]string, error) {
 		}
 	}
 
-	if value == "" {
+	if eqFound && f.isBoolFlag && value == "" {
+		value = "true" // `--debug=` amounts to `--debug=true`
+	}
+
+	if value == "" && !eqFound {
 		switch {
 		case f.isBoolFlag:
-			value = "true" // `-b` or `--foo` default to true
+			value = "true" // `--foo` defaults to true
 			if len(args) > 0 {
 				if _, err := strconv.ParseBool(args[0]); err == nil {
-					value = args[0] // `-b true` or `--foo false` should also work
+					value = args[0] // `--foo false` should also work
 					args = args[1:]
 				}
 			}
```

There is also something strange with the code (the code that already exists)

Let's assume --foo expects a string

--foo leads to an error
--foo true leads to "true"
--foo=1 leads to "1"
--foo=false leads to "false"
--foo= leads to ""

But if --foo is a boolean

--foo false
--foo whatever
--foo 0
--foo 2
--foo=2

Leads to this

--foo=false
--foo whatever
--foo=false
--foo=true 2
an error

This behavior is strange to me, but I'm unsure how other libraries parsing flags do

---

The difference between `--foo 2` and `--foo=2` is...not great. That is probably why `flag` in Go's standard library [restricts the form `--flag arg` to non-boolean flags](https://cs.opensource.google/go/go/+/refs/tags/go1.23.4:src/flag/flag.go;l=54).

---

Exactly

---

I would be in favor of that here as well, but I think @peterbourgon wants to keep boolean parsing as is. (Also, that would probably be a very breaking API change now. Don't know how that plays with v4 being in beta.)

---

@ccoVeille As a side note, `flag` in the standard library has a parallel inconsistency but for string flags.

```go
	fs := flag.NewFlagSet("whatever", flag.ContinueOnError)
	cfg := struct {
		config  string
		// ...
	}{}
	fs.StringVar(&cfg.config, "config", "", "Use this configuration file")
	// ...
```

```shell
whatever -config=foo   # No error; config is set to "foo"
whatever -config foo   # No error; config is set to "foo"
whatever -config ""    # No error; config is set to ""
whatever -config=""    # No error; config is set to ""
whatever -config       # Error, namely "flag needs an argument: -config"
whatever -config=      # No error; config is set to ""
```

I think that the last two should return the same error, but they do not. I'm guessing this is because the parser would need to do extra work (not much but some) to detect the difference between `-config=` and `-config=""`. I doubt anything can be done about it now (breaking API change), but I think it was a mistake.

---

For me, the behavior of the stdlib is fine here. I might look a bit odd, but I'm fine with it.

It has its logic.

About ff lib, I don't know really. You are fixing a bug with the boolean flag after all, so fixing the bug is somehow already breaking something that was broken :smile:, by fixing it.

While your fix is fine, I think the issue of removing the random behavior of boolean flag with a parameter should be considered, at least to have a library that behave like other libraries.

I don't think it was intended, I would remove it. But, except that your PR is fine, you are fixing the behavior with `--foo= ` so nothing about `--foo 0`

So for me it can be merged as is

---

For record, I randomly found a discussion about the exact same issue with boolean flag

- https://github.com/spf13/pflag/issues/410